proposals/XXXX-build-time-constant-values.md

```diff
@@ -0,0 +1,188 @@
+
+## Detailed Design
+
+### Property `@const` attribute
+
+A stored property on a `struct` or a `class` can be marked with a `@const` attribute to indicate that its value is known at compile-time. 
+
+```
+struct Foo {
+  @const let title: String = "foo"
+}
+```
+
+Similarly to Implicitly-Unwrapped Optionals, the mental model for semantics of this attribute is that it is a flag on the declaration that  guarantees that the compiler is able to know its value as shared by all instance of the type. For now, `@const let` and `@const static let` are equivalent in what information the `@const` attribute conveys to the compiler.(*AC: Worth mentioning here that in the future,* `@const let` *properties will gain an ability to be initialized with compile-time expressions*). Unlike a plain `static let` stored property, a `@const static let` property does not need to refer to a location in memory shared by all instances of the type, and can be elided by the compiler entirely. Default-initializing a `@const` property with a runtime value or not default-initializing it at all results in an compilation error.
+
+### Parameter `@const` attribute
+
+A function parameter can be marked with a `@const`  keyword to indicate that values passed to this parameter at the call-site must be compile-time-known values.
+
+```
+func foo(input: @**const** Int) {...}
+```
+
+### Supported Types
+
+The requirement that values of `@const` properties and parameters be known at compile-time restricts the allowable types for such declarations. The current scope of the proposal includes:
+
+* Integer and floating-point literals.
+* String literals (excluding interpolated strings).
+* Enum cases with no associated values.
+* Array literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+* Dictionary literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+
+This list may expand in the future to include more literal-value kinds or potential new compile-time valued constructs.
+
+### Protocol `@const` property requirement 
+
+A protocol author may require conforming types to default initialize a given property with a compile-time-known value by specifying it as `@const static` in the protocol definition. For example:
+
+```
+protocol NeedsConstGreeting {
+    @const static let greeting: String
+}
+```
+
+If a conforming type initializes `greeting` with something other than a compile-time-known value, a compilation error is produced:
+
+```
+struct Foo: NeedsConstGreeting {
+  // 👍
+  static let greeting = "Hello, Foo"
+}
+
+struct Bar: NeedsConstGreeting {
+  // error: 'greeting' must be initialized with a const value
+  static let greeting = "\(Bool.random ? "Hello" : "Goodbye"), Bar"
+}
+```
+
+## Motivating Example Use-Cases
+
+### Declarative Package Manifests that do not require execution
```

Comment:
This section heading differs from the other because it cites the example, not the general design goal. Suggest renaming it to "Facilitate Compile-time Extraction of Values".